src/packet/public_key_encrypted_session_key.rs

```diff
@@ -13,95 +13,252 @@ use crate::crypto::sym::SymmetricKeyAlgorithm;
         data[1..=len].copy_from_slice(session_key);
 
-        // and appended a checksum
-        BigEndian::write_u16(
-            &mut data[len + 1..],
-            checksum::calculate_simple(session_key),
-        );
+        // Append a checksum, except for x25519/x448
+        // FIXME: factor this difference out and up?
+        match pkey.public_params() {
+            PublicParams::X25519 { .. } | PublicParams::X448 { .. } => {}
+            _ => {
+                // and appended a checksum
+                data.extend_from_slice(&checksum::calculate_simple(session_key).to_be_bytes())
+            }
+        }
 
-        let mpis = pkey.encrypt(rng, &data)?;
+        let values = pkey.encrypt(rng, &data, EskType::V3_4)?;
 
-        Ok(PublicKeyEncryptedSessionKey {
+        Ok(PublicKeyEncryptedSessionKey::V3 {
             packet_version: Default::default(),
-            version: 3,
             id: pkey.key_id(),
-            algorithm: pkey.algorithm(),
-            mpis,
+            pk_algo: pkey.algorithm(),
+            values,
         })
     }
 
-    pub fn id(&self) -> &KeyId {
-        &self.id
+    /// Encrypts the given session key to `pkey` as a v6 pkesk.
+    pub fn from_session_key_v6<R: CryptoRng + Rng>(
+        rng: R,
+        session_key: &[u8],
+        pkey: &impl PublicKeyTrait,
+    ) -> Result<Self> {
+        // the session key
+        let mut data = session_key.to_vec();
+
+        // Append a checksum, except for x25519/x448
+        // FIXME: factor this difference out and up?
+        match pkey.public_params() {
+            PublicParams::X25519 { .. } | PublicParams::X448 { .. } => {}
+            _ => data.extend_from_slice(&checksum::calculate_simple(session_key).to_be_bytes()),
+        }
+
+        let values = pkey.encrypt(rng, &data, EskType::V6)?;
+
+        Ok(PublicKeyEncryptedSessionKey::V6 {
+            packet_version: Default::default(),
+            fingerprint: Some(pkey.fingerprint()),
+            pk_algo: pkey.algorithm(),
+            values,
+        })
     }
 
-    pub fn mpis(&self) -> &[Mpi] {
-        &self.mpis
+    /// Check if a Key matches with this PKESK's target
+    /// - for v3: is PKESK key id the wildcard, or does it match the key id of `pkey`?
+    /// - for v6: is PKESK fingerprint the wildcard (represented as `None`), or does it match the fingerprint of `pkey`?
+    pub fn match_identity(&self, pkey: &impl PublicKeyTrait) -> bool {
+        match self {
+            Self::V3 { id, .. } => id.is_wildcard() || (id == &pkey.key_id()),
+            Self::V6 { fingerprint, .. } => {
+                if let Some(fingerprint) = fingerprint {
+                    fingerprint == &pkey.fingerprint()
+                } else {
+                    true // wildcard always matches
+                }
+            }
+            _ => false,
+        }
     }
 
-    pub fn packet_version(&self) -> Version {
-        self.packet_version
+    pub fn id(&self) -> Result<&KeyId> {
```

Comment:
docs missing for this and following methods

Reply:
right, wrote docs in bb714ecdbe4d5d0ce7fb169a651969cf8257ab6b